File: Programmers/조이스틱/Woon/조이스틱.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int left_count(string name, int index) {

	int size = name.size();
	int count = 0;
	while (true) {


		if (count == size - 1) {
			break;
		}

		index--;
		if (index < 0) {
			index = size - 1;
		}

		if (name[index] == 'A') {
			count++;
		}
		else {
			count++;
			break;
		}
	}

	return count;
}

int right_count(string name, int index) {

	int size = name.size();
	int count = 0;
	while (true) {

		if (count == size - 1) {
			break;
		}

		index++;
		if (index > size-1) {
			index = 0;
		}

		if (name[index] == 'A') {
			count++;
		}
		else {
			count++;
			break;
		}
	}

	return count;
}

bool a_check(string name) {
	int size = name.size();
	bool check = true;
	for (int i = 0; i < size; i++) {
		if (name[i] != 'A') {
			check = false;
			break;
		}
	}
	return check;
}


int solution(string name) {
	int answer = 0;

	int size = name.size();
	int count = 0;
	for (int i = 0; i < size; i++) {
		int x = name[i] - 65;

		if (x < 13) {
			count = count + x;
		}
		else {
			count = count + (26 - x);
		}
	}

	int left = 0;
	int right = 0;
	int index = 0;
	int move = 0;

	while (true) {
		
	
		if (index < 0) {
			index = size- 1;
		}
		if (index > size - 1) {
			index = 0;
		}

		if (name[index] == 'A') {
			left = left_count(name, index);
			right = right_count(name, index);
		}
		else {
			name.replace(index, 1, "A");
			left = left_count(name, index);
			right= right_count(name, index);
		}

		if (left < right) {
			index--;
		}
		else {
			index++;
		}

		move++;
		if (a_check(name) == true) {
			break;
		}
	}

	answer = move + count - 1;


	return answer;
}
int main() {


	string name = "JAN";
	cout << solution(name);

	return 0;	
}

File: Programmers/조이스틱/Hong/조이스틱.java
package com.company;

import java.util.*;

class Solution {
    public int solution(String name) {
        int answer = 0;
        int size = name.length();
        int index = 0;
        int left = index, right = index;
        StringBuffer res = new StringBuffer();
        StringBuffer init = new StringBuffer();
        int leftIndex=0;
        int rightIndex=0;

        res.append(name);

        for(int i=0; i<size; i++){
            init.append('A');
        }


        while(true) {
            if(right <= left){
                answer += right;
                index = rightIndex;
            }
            else{
                answer += left;
                index = leftIndex;
            }

            char ch = name.charAt(index);
            if(ch != 'A'){
                answer += Math.min(ch - 'A', 'Z' - ch + 1);
                res.replace(index,index+1,"A");
            }

            if(init.toString().equals(res.toString())){
                break;
            }

            leftIndex = index;
            rightIndex = index;
            for(int i=1; i<size; i++){
                if(rightIndex == size - 1){
                    rightIndex = 0;
                }
                else{
                    rightIndex++;
                }

                if(res.charAt(rightIndex)!='A'){
                    right = i;
                    break;
                }
            }

            for(int i=1; i<size; i++){
                if(leftIndex == 0){
                    leftIndex = size - 1;
                }
                else{
                    leftIndex--;
                }

                if(res.charAt(leftIndex)!='A'){
                    left = i;
                    break;
                }
            }
        }
        return answer;
    }
}

class Main {

    public static void main(String[] args) {

        //String name = "JAZ";
        //String name = "JEROEN";
        //String name = "AAAB";

        //9
        String name = "BBBAAAB";

        //11
        //String name = "ABABAAAAABA";

        //String name = "ABAA";

        Solution solution = new Solution();
        System.out.println(solution.solution(name));
    }
}
